segmentation/utils/synthetic_spatial_datagen.py

```diff
@@ -138,35 +139,158 @@ def point_init_dist_matrix(size_img=(1024, 1024), num_A=100, num_B=100, num_C=10
     # since label only takes a binary matrix
     # we pass the result through the unique function to eliminate any possibility of duplicate points
     # appearing within any of these arrays
-    a_points = np.unique(np.random.multivariate_normal(a_mean, a_cov, num_A).astype(np.int16), axis=0)
-    b_points = np.unique(np.random.multivariate_normal(b_mean, b_cov, num_B).astype(np.int16), axis=0)
-    c_points = np.unique(np.random.multivariate_normal(c_mean, c_cov, num_C).astype(np.int16), axis=0)
+    # because of the small possiblity of generating a list of points that are duplicate, we use a while loop
+    # to regenerate a new set of points for a, b, and c in case of failure
+    non_dup_array = False
 
-    # this ensures that we only keep the points that are not duplicate across different cell types
-    points, counts = np.unique(np.concatenate((a_points, b_points, c_points), axis=0), axis=0, return_counts=True)
-    non_dup_points = points[counts == 1]
+    while not non_dup_array:
+        a_points = np.abs(np.unique(np.random.multivariate_normal(a_mean, a_cov, num_A).astype(numpy_type), axis=0))
+        b_points = np.abs(np.unique(np.random.multivariate_normal(b_mean, b_cov, num_B).astype(numpy_type), axis=0))
+        c_points = np.abs(np.unique(np.random.multivariate_normal(c_mean, c_cov, num_C).astype(numpy_type), axis=0))
 
-    # an astronomically unlikely error, would occur if for some reason all of the points were labelled as duplicate
-    if len(non_dup_points) == 0:
-        raise ValueError("Bad run: no unique points generated. Try again, will work next time.")
+        total_points = np.concatenate((a_points, b_points, c_points), axis=0)
 
-    point_x_coords, point_y_coords = zip(*non_dup_points)
+        # this ensures that we only keep the points that are not duplicate across different cell types
+        points, counts = np.unique(total_points, axis=0, return_counts=True)
+        dup_points = points[counts > 1]
 
-    # generate the binary matrix to pass into label_map
-    binary_mat = np.zeros(size_img)
-    binary_mat[point_x_coords, point_y_coords] = True
+        # check that there are no duplicate points and that the ranges of the random points generated fall
+        # within range, otherwise we're regenerating the set of points with a new random seed
+        if len(dup_points) == 0 and len(total_points[:, 0][total_points[:, 0] > size_img[0]]) == 0 and \
+           len(total_points[:, 0][total_points[:, 0] > size_img[0]]) == 0 and \
+           len(total_points[:, 0][total_points[:, 0] < 0]) == 0 and \
+           len(total_points[:, 1][total_points[:, 1] > size_img[1]]) == 0 and \
+           len(total_points[:, 1][total_points[:, 1] < 0]) == 0:
 
-    # generate the label matrix for the image now
-    label_mat = label(binary_mat)
+            non_dup_array = True
+        else:
+            if seed:
+                np.random.seed(seed + 1)
```

Comment:
I don't really see the need for the big while loop here.  If we don't really need exactly `num_X` cells of type X, then just removing duplicates (already done with `np.unique`) and removing out of bounds points is fine.  Also, using `np.abs` here makes the distributions kinda wonky (reflections off of x & y axis), so just remove any negative points.  

If exactly `num_X` cells of type X are needed, this is extra extra arduous, and currently isn't happening (i.e `np.unique` is called when initing `a_points`, which removes duplicates).  

If the former is the case, `num_X` could be renamed to `max_num_X` or something like that

Reply:
Looking back, I realize the main motivation for the while loop was to avoid the hassle of having to redo Jay's testing function, specifically make_expression_matrix. If you leave points out, then Jay's tests are no longer valid because you don't have the same number of cell centroids in the distance matrix as you do in the expression matrix.

However, with Noah's point about not needing to test this function, it's probably safe to leave points out. Definitely makes life easier!

All that hard work to make things just right and the simple solution ended up working best in the end 😂.

Reply:
The other point of this PR is to see if there are any changes that need to be made to the way Jay implemented these functions. If it would make our life easier to change the assumptions about data format, indexing, etc, those are all on the table for this. Don't feel like you have to code around whatever initial choices were made, nothing is set in stone here, especially if it will reduce the complexity of the stuff we have to write on top.